homeassistant/components/binary_sensor/flic.py

```diff
@@ -0,0 +1,240 @@
+"""Contains functionality to use flic buttons as a binary sensor."""
+import asyncio
+import logging
+from datetime import datetime
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.const import CONF_HOST, CONF_PORT, EVENT_HOMEASSISTANT_STOP
+from homeassistant.components.binary_sensor import (
+    BinarySensorDevice, PLATFORM_SCHEMA)
+from homeassistant.util.async import (
+    fire_coroutine_threadsafe, run_callback_threadsafe)
+
+
+REQUIREMENTS = ['https://github.com/soldag/pyflic/archive/0.4.zip#pyflic==0.4']
+
+_LOGGER = logging.getLogger(__name__)
+
+
+CONF_AUTO_SCAN = "auto_scan"
+CONF_THRESHOLD_DOUBLE_CLICK = "double_click_threshold"
+CONF_THRESHOLD_LONG_CLICK = "long_click_threshold"
+
+EVENT_FLIC_SINGLE_CLICK = "flic_single_click"
+EVENT_FLIC_LONG_CLICK = "flic_long_click"
+EVENT_FLIC_DOUBLE_CLICK = "flic_double_click"
+
+
+# Validation of the user's configuration
+THRESHOLDS_SCHEMA = vol.Schema({
+})
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_HOST, default='localhost'): cv.string,
+    vol.Optional(CONF_PORT, default=5551): cv.port,
+    vol.Optional(CONF_AUTO_SCAN, default=True): cv.boolean,
+    vol.Optional(CONF_THRESHOLD_LONG_CLICK, default=0.5): float,
+    vol.Optional(CONF_THRESHOLD_DOUBLE_CLICK, default=0.3): float
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_entities,
+                         discovery_info=None):
+    """Setup the flic platform."""
+    import pyflic
+
+    # Get event loop
+    loop = asyncio.get_event_loop()
+
+    # Initialize main flic client responsible for
+    # connecting to buttons and retrieving events
+    host = config.get(CONF_HOST)
+    port = config.get(CONF_PORT)
+    try:
+        main_client = pyflic.FlicClient(host, port)
+    except ConnectionRefusedError:
+        _LOGGER.error("Failed to connect to flic server.")
+        return
+
+    def new_button_callback(address):
+        """Setup newly verified button as device in home assistant."""
+        hass.loop.create_task(async_setup_button(hass, config,
+                                                 async_add_entities,
+                                                 main_client, address))
+
+    main_client.on_new_verified_button = new_button_callback
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, main_client.close)
+    loop.run_in_executor(None, main_client.handle_events)
```

Comment:
No, `FlicClient.handle_events()` has to run the entire time in background. It is a main loop of the pyflic lib that never exits, unless the socket to the flic service is closed.